driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -149,6 +156,7 @@ public InternalConnection get(final long timeoutValue, final TimeUnit timeUnit)
         connectionPoolListener.connectionCheckOutStarted(new ConnectionCheckOutStartedEvent(serverId));
         Timeout timeout = Timeout.startNow(timeoutValue, timeUnit);
         try {
+            stateAndGeneration.throwIfClosedOrPaused();
```

Comment:
No there is no such guarantee, even though this may not be the only "is paused" check when checking out (there are also checks in the `ConcurrentPool` and in `OpenConcurrencyLimiter.acquirePermitOrGetAvailableOpenedConnection`). The reason there is no such guarantee is that while the requirement may be formulated from the standpoint of internals of `DefaultConnectionPool`, it cannot be formulated from the standpoint of a user of `DefaultConnectionPool`:

1. In the context of this matter in question, a pool provides two operations to its users: checkout (`get`/`getAsync`) and pause (`invalidate`), which are allowed to be called concurrently with respect to each other. We will consider executions in which the pool is ready (situations in which it is paused are not interesting for the purpose of this discussion).
2. If pause completes before (in `hb`) checkout is invoked, then a user expects the checkout to fail with `PoolClearedException` and the pool is guaranteed to produce `PoolClearedException` (unless the pool is marked ready concurrently with the checkout operation, but such executions seem to not be part of the question).
3. If checkout completes (whether synchronously or asynchronously) before pause is invoked, then a user expects the checkout to not result in `PoolClearedException` and the pool is guaranteed to not throw `PoolClearedException`.
4. If neither 2 nor 3, then operations pause and checkout are concurrent by definition, and a user may not have an expectation on whether checkout produces `PoolClearedException` or not. For the pool this means that it is free to either produce a connection or produce `PoolClearedException`.

Reply:
Got it.  Thanks.